text/0000-quick-debug-macro.md

```diff
@@ -0,0 +1,183 @@
+- Feature Name: quick_debug_macro
+- Start Date: 2017-10-13
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Adds a macro `dbg!(expr)` for quick and dirty `Debug`:ing of an expression to the terminal. The macro evaluates the expression, prints it to `STDERR`, and finally yields `expr`. On release builds, the macro is the identity function and has no side effects. The macro is added to the prelude of the standard library.
+
+# Motivation
+[motivation]: #motivation
+
+The motivation to add this new `dbg!` macro is two-fold.
+
+## For aspiring rustaceans
+
+One of the often asked questions is how to print out variables to the terminal.
+Delaying the need to explain formatting arguments in the statement `println!("{:?}", expr);` can help aspiring rustaceans to quickly learn the language. With `dbg!(expr);` there is no longer such need, which can be delayed until the developer actually cares about the format of the output and not just what value the expression evaluates to.
+
+## For experienced developers
+
+By using `dbg!(expr);`, the burden of a common papercut: writing `println!("{:?}", expr);` every time you want to see the evaluted-to value of an expression, can be significantly reduced. The developer no longer has to remember the formatting args and has to type significantly less (12 characters to be exact).
+
+The shortness is also beneficial when asking `playbot-mini` on #rust@irc.mozilla.org to evaluate and print an expression.
```

Comment:
I think IRC bots would rather a very different output format from what's described currently:
* The `[DEBUGGING, src/main.rs:1:12]:` is clearly of no value there
* Most bots only keep one or two lines (to avoid spamming), so `{:#?}` is worse than `{:?}`
* The expression is one the line right above, so isn't really worth showing again

`playbot` (RIP) just used `fn show<T:Debug>(x: T) -> T { print!("{:?}", x); x }` which worked great for it.

Reply:
I buy this =) I'll remove the argument from the RFC.